server/src/lib/initializer.rs

```diff
@@ -241,109 +256,170 @@ impl<'a> MachineInitializer<'a> {
+                    ),
+                )
+            })?;
+            let be_info = backend.info();
+            match device_spec.kind {
+                StorageDeviceKind::Virtio => {
+                    let vioblk = virtio::PciVirtioBlock::new(0x100, be_info);
+                    let id =
+                        self.inv.register_instance(&vioblk, bdf.to_string())?;
+                    let _ = self.inv.register_child(child, id).unwrap();
+                    backend.attach(vioblk.clone(), self.disp)?;
+                    chipset.device().pci_attach(bdf, vioblk);
+                }
+                StorageDeviceKind::Nvme => {
+                    let nvme = nvme::PciNvme::create(name.to_string(), be_info);
+                    let id =
+                        self.inv.register_instance(&nvme, bdf.to_string())?;
+                    let _ = self.inv.register_child(child, id).unwrap();
+                    backend.attach(nvme.clone(), self.disp)?;
+                    chipset.device().pci_attach(bdf, nvme);
+                }
+            };
+        }
         Ok(())
     }
 
-    pub fn initialize_crucible(
+    pub fn initialize_network_devices(
         &self,
         chipset: &RegisteredChipset,
-        disk: &propolis_client::api::DiskRequest,
-        bdf: pci::Bdf,
     ) -> Result<(), Error> {
-        info!(self.log, "Creating Crucible disk from {:#?}", disk);
-        let be = propolis::block::CrucibleBackend::create(
-            disk.gen,
-            disk.volume_construction_request.clone(),
-            disk.read_only,
-        )?;
-
-        info!(self.log, "Creating ChildRegister");
-        let creg = ChildRegister::new(&be, Some(be.get_uuid()?.to_string()));
-
-        match disk.device.as_ref() {
-            "virtio" => {
-                info!(self.log, "Calling initialize_virtio_block");
-                self.initialize_virtio_block(chipset, bdf, be, creg)
-            }
-            "nvme" => {
-                info!(self.log, "Calling initialize_nvme_block");
-                self.initialize_nvme_block(
-                    chipset,
-                    bdf,
-                    disk.name.clone(),
-                    be,
-                    creg,
+        for (name, vnic_spec) in &self.spec.network_devices {
+            info!(self.log, "Creating vNIC {}", name);
+            let backend_spec = self
+                .spec
+                .network_backends
+                .get(&vnic_spec.backend_name)
+                .ok_or_else(|| {
+                    Error::new(
+                        ErrorKind::InvalidInput,
+                        format!(
+                            "Backend {} not found for vNIC {}",
+                            vnic_spec.backend_name, name
+                        ),
+                    )
+                })?;
+            let bdf = pci::Bdf::new(
+                vnic_spec.pci_path.0,
+                vnic_spec.pci_path.1,
+                vnic_spec.pci_path.2,
```

Comment:
Ok, looks like we do this at least a couple times so maybe add a From impl so we don't need to repeat as much.

Reply:
This was mildly annoying at first because `propolis-server` is the only crate that imports both the `PciPath` and `Bdf` types, but because both are external types, it can't `impl From<PciPath> for Bdf`. Rather than try to resolve that in the server itself, I added a small `types` crate to hold `PciPath` so that the Propolis lib crate can include it and define the From impl. 

If there's a better way to do this, let me know--this seemed simpler than adding a wrapper newtype in the server but I'm not at all sure I didn't overlook an even better approach.

In the longer term we could probably even move the `Bdf` struct out of the lib and into `types` and do all the device/function validation on that single type. But I'm inclined to let the dust settle on this change before doing much more refactoring with it.